### PR TITLE
Use binary represention for relations in qsieve

### PR DESCRIPTION
**Pull request by albinahlback**
Finally got it to work (I think).

Edit: Perhaps Windows machines needs to use open the files in binary mode.

### CODE REVIEW

File: src/qsieve/large_prime_variant.c [hunk 1]
```diff
@@ -206,52 +251,71 @@ void qsieve_add_to_hashtable(qs_t qs_inf, mp_limb_t prime)
...
-        str = next;
+    /* Get Y */
+    fmpz_init(rel.Y);
+    if (FLINT_ABS(Ysz) <= 1)
+    {
+        mp_limb_t abslimb = 0;
 
-        while (isspace(*str))
-            str++;
+        fread(&abslimb, sizeof(mp_limb_t), 1, (FILE *) qs_inf->siqs);
```
Comment (fredrik-johansson) [inline src/qsieve/large_prime_variant.c:291]: Why not just use ``fmpz_set_ui`` for this block of code?

**Comment by fredrik-johansson** [comment]
This is good (I just had one small nitpick from looking at the code), and should hopefully make it easier to get rid of the file management altogether.

I wonder if the default test code is enough or if we should do some more stress testing with large factorisations (just before merging before this, not necessarily as part of the test suite)?

**Comment by albinahlback** [comment]
> This is good (I just had one small nitpick from looking at the code), and should hopefully make it easier to get rid of the file management altogether.

Yeah, I definitely think this is on the right path.

> I wonder if the default test code is enough or if we should do some more stress testing with large factorisations (just before merging before this, not necessarily as part of the test suite)?

Sounds good. Could you implement some more extensive tests? I have no idea of how the module actually works.

**Comment by fredrik-johansson** [comment]
> Sounds good. Could you implement some more extensive tests? I have no idea of how the module actually works.

You could just try to generate some random big semiprimes:

    sage: random_prime(10^30) * random_prime(10^30)
    353002404869087467911973946894285937737279398804756390360397
    sage: random_prime(10^35) * random_prime(10^35)
    3276353236778190952987709831996872642885327175178186820567627852001759
    sage: random_prime(10^40) * random_prime(10^40)
    19714781844001166488455486606377739043994798987566671863342247690296820624767507

Then run the example factoring program and make sure it succeeds:

    build/examples/factor_integer -threads 8 -timing 353002404869087467911973946894285937737279398804756390360397
    build/examples/factor_integer -threads 8 -timing 3276353236778190952987709831996872642885327175178186820567627852001759
    build/examples/factor_integer -threads 8 -timing 19714781844001166488455486606377739043994798987566671863342247690296820624767507

80 digits should take a couple of minutes which is probably a good enough stress test.

While you're at it, you could maybe also change the debug ``printf`` calls to ``flint_printf`` in this module so that it works if one sets ``#define QS_DEBUG (64|128)`` in ``qsieve.h``.

**Comment by albinahlback** [comment]
Seems to work fine:
```
~/c/flint (qsieve_binary) $ build/examples/factor_integer -threads 4 -timing 353002404869087467911973946894285937737279398804756390360397
cpu/wall(s): 10.006 3.71
virt/peak/res/peak(MB): 246.72 267.86 16.48 32.11
353002404869087467911973946894285937737279398804756390360397 =
587435218207196569871936744563 * 600921418954793771220512089919

~/c/flint (qsieve_binary) $ build/examples/factor_integer -threads 4 -timing 3276353236778190952987709831996872642885327175178186820567627852001759
cpu/wall(s): 132.761 40.156
virt/peak/res/peak(MB): 276.21 338.68 46.37 87.27
3276353236778190952987709831996872642885327175178186820567627852001759 =
46482866437252613754572804490573859 * 70485180624584583358627850098638101

~/c/flint (qsieve_binary) $ build/examples/factor_integer -threads 4 -timing 19714781844001166488455486606377739043994798987566671863342247690296820624767507
cpu/wall(s): 857.685 282.414
virt/peak/res/peak(MB): 290.09 367.15 64.18 113.40
19714781844001166488455486606377739043994798987566671863342247690296820624767507 =
2673983399403925307863620624678723438299 * 7372813850824920691267112915018383684393
```

**Comment by fredrik-johansson** [comment]
Surely

    #define QS_DEBUG 0 /* level of debug information printed, 0 = none */

should be kept?

**Comment by albinahlback** [comment]
> Surely
> 
> ```
> #define QS_DEBUG 0 /* level of debug information printed, 0 = none */
> ```
> 
> should be kept?

No, I think this is something that should be put into `CPPFLAGS`. It should not bloat the namespace of the header.

**Comment by fredrik-johansson** [comment]
Having that in the header file makes it self-documenting though. And this is clearly a debug feature internal to the qsieve module.

**Comment by albinahlback** [comment]
I think this should be good. Will merge if the tests passes, unless you say something else.

**Comment by fredrik-johansson** [comment]
Thumbs up!

**Pull request merged**